CAT thread links

ianrowan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Link to Issue
Closes: #5040 

## Description of Changes
- Adds the ability to link a CAT for display in sidebar or in-line on threads 
- Adds the display option configuration to enable the sidebar vs inline
- Introduces a new `view_template` style for showing in forms and cards
<img width="362" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/36428666/68c37745-5135-424f-8600-a3e70748f495">


## Test Plan
- with CAT flag on create a thread in a community with active CATs
- add the CAT via the card and modal
- ensure it displays in sidebar by default
- test toggle of sidebar/default 
- attempt a transaction and ensure metamask pops up
- ensure regular template UX is not changed

## Other Considerations
<!--- Follow-up tickets, breaking changes, etc -->
- for context the UX/usefulness of this feature is for threads that point to a specific contract action(ie vote/propose/delegate). This allows the ability to create on-chain interactive threads with two display options 
- currently uses the CAT feature flag as its assumed this feature can be enabled if CAT is enabled. Can be updated to more granular flag if necessary

masvelio commented on packages/commonwealth/client/scripts/views/components/TemplateActionSelector/templateActionSelector.tsx:
```diff
@@ -0,0 +1,109 @@
+export const TemplateSelector = ({
+  onSelect,
+  tempTemplates,
+  isOpen,
```
'isOpen' is defined but never used

ianrowan replied:
✅

masvelio commented on packages/commonwealth/client/scripts/views/components/TemplateActionSelector/templateActionSelector.tsx:
```diff
@@ -0,0 +1,109 @@
+}: TemplateSelectorProps) => {
+  // Add a new state for contracts
+  const [loading, setLoading] = useState(false);
+  const [fetched, setFetched] = useState(false);
```
'fetched' is assigned a value but never used.
Probably you dont need this line at all

masvelio commented on packages/commonwealth/client/scripts/views/components/TemplateActionSelector/templateActionSelector.tsx:
```diff
@@ -0,0 +1,109 @@
+  const templates = useMemo(() => {
+    if (!searchTerm.length) return allTemplates;
+    else {
+      console.log(allTemplates);
```
remove console log

masvelio commented on packages/commonwealth/client/scripts/views/components/TemplateActionSelector/templateActionSelector.tsx:
```diff
@@ -0,0 +1,109 @@
+      />
+    </div>
+  );
+};
```
1. file should be Capitalised
2. File name and component name should be the same so either `templateActionSelector` or `TemplateSelector`

masvelio commented on packages/commonwealth/client/scripts/views/components/TemplateActionSelector/templateActionSelectorItem.tsx:
```diff
@@ -0,0 +1,33 @@
+      </div>
+    </div>
+  );
+};
```
1. file should be Capitalised
2. File name and component name should be the same so either templateActionSelectorItem or TemplateSelectorItem

masvelio commented on packages/commonwealth/client/scripts/views/modals/template_action_modal.tsx:
```diff
@@ -0,0 +1,183 @@
+      </div>
+    </div>
+  );
+};
```
1. Capitalise and CamelCase for file name

masvelio commented on packages/commonwealth/client/scripts/views/modals/template_action_modal.tsx:
```diff
@@ -0,0 +1,183 @@
+import { filterLinks, getAddedAndDeleted } from '../../helpers/threads';
+import { CWIconButton } from '../components/component_kit/cw_icon_button';
+import { TemplateSelector } from '../components/TemplateActionSelector';
+import { CWButton } from '../components/component_kit/cw_button';
```
use CWButton from `new_designs/cw_button`

masvelio commented on packages/commonwealth/client/scripts/views/modals/template_action_modal.tsx:
```diff
@@ -0,0 +1,183 @@
+    chainId: app.activeChainId(),
+    threadId: thread.id,
+  });
+  // Fetch contracts
```
this comment is redundant

masvelio commented on packages/commonwealth/client/scripts/views/modals/template_action_modal.tsx:
```diff
@@ -0,0 +1,183 @@
+  const fetchContracts = async () => {
+    setLoading(true);
+    const contractsInStore = app.contracts.getCommunityContracts();
+    console.log(contractsInStore);
```
remove console logs

masvelio commented on packages/commonwealth/client/scripts/views/modals/template_action_modal.tsx:
```diff
@@ -0,0 +1,183 @@
+    );
+  };
+
+  const getContractAndCct = (identifier, _contracts) => {
```
'_contracts' is defined but never used

masvelio commented on packages/commonwealth/client/scripts/views/modals/template_action_modal.tsx:
```diff
@@ -0,0 +1,183 @@
+        'identifier'
+      );
+
+      console.log('toAdd:', toAdd);
+      console.log('toDelete:', toDelete);
```
remove console logs

masvelio commented on packages/commonwealth/client/scripts/views/pages/view_template/view_template.tsx:
```diff
@@ -69,13 +76,20 @@ const ViewTemplatePage = () => {
       return cct.cctmd.slug === slug || cct.cctmd.slug === `/${slug}`;
     });
-
+    console.log(contractInStore, slug, contractInStore.ccts, templateMetadata);
```
remove console log

masvelio commented on packages/commonwealth/client/styles/modals/template_action_modal.scss:
```diff
@@ -0,0 +1,105 @@
+  }
+}
+
+.template-item {
+  display: flex;
+  flex-direction: row;
+  padding: 1rem;
+  border: 1px solid #ccc;
+  border-radius: 4px;
+  cursor: pointer;
+
+  &:hover {
+    background-color: #f0f0f0;
+  }
+
+  .selected {
+    min-width: 50px;
+    display: flex;
+    justify-content: center;
+    align-items: center;
+
+    svg {
+      fill: $black;
+      width: 18px;
+    }
+  }
+
+  .template-nickname {
+    font-weight: bold;
+    margin-bottom: 0.5rem;
+  }
+
+  .template-contract-address {
+    font-size: 0.8rem;
+    color: #888;
+  }
+}
```
move this to separate file because you use it in TemplateSelectorItem component

masvelio commented on packages/commonwealth/client/styles/modals/template_action_modal.scss:
```diff
@@ -0,0 +1,105 @@
+  }
+}
+
+.TemplateSelector {
```
same here, move it to seperate scss file

masvelio commented on packages/commonwealth/client/scripts/views/pages/view_thread/viewTemplateFormCard.tsx:
```diff
@@ -0,0 +1,24 @@
+      }
+    />
+  );
+};
```
ViewTemplateFormCard.tsx

masvelio commented on packages/commonwealth/client/scripts/views/pages/view_thread/templateActionCard.tsx:
```diff
@@ -0,0 +1,122 @@
+      />
+    </div>
+  );
+};
```
TemplateActionCard.tsx

masvelio reviewed (commented):
Please make make sure that the naming of the tsx, scss files names and also component names are proper.

I am getting infinite cow without any information when I open up this modal. It failed to get templates.

![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/324042d2-454a-4b1e-9d2d-59880b9501e6)

ianrowan commented:
> Please make make sure that the naming of the tsx, scss files names and also component names are proper.
> 
> I am getting infinite cow without any information when I open up this modal. It failed to get templates.

Corrected all of the file name issues and comments. As for this infinite cow - this was occurring for contracts set up before templates went in that do not have an abi in the database(ie dydx) you can try on another community like 1inch and it should work(although likely no templates) 

To address the infinite cow I added some error handling to fail more gracefully and provide a no template message 

@masvelio

masvelio commented on packages/commonwealth/client/scripts/views/pages/view_template/view_template.tsx:
```diff
@@ -48,9 +48,16 @@ type Json = {
+const ViewTemplatePage = (formData?: ViewTemplateFormProps) => {
   const navigate = useCommonNavigate();
-  const params = useParams();
+  const params = formData;
```
remove useParams from import in l. 8

ianrowan commented:
> ![image](https://user-images.githubusercontent.com/14819225/268275334-db977505-7cb8-40bd-a2bd-397ea922accc.png)
> 
> Those files should also be in PascalCase. Basically all files on front end should follow this pattern. Let's not create snake_case file anymore, please.
> 
> Apart from that, code looks good! Thanks for the updates!

Fixed those as well

jnaviask reviewed (approved):
Should be OK to go in for QA. Can merge pending Session Keys.
